tests/test_bounce_integral.py

```diff
@@ -0,0 +1,783 @@
+    """Test vectorized computation of polynomial derivative."""
+    quintic = 6
+    c = np.arange(-18, 18).reshape(quintic, 3, -1) * np.pi
+    # make sure broadcasting won't hide error in implementation
+    assert np.unique(c.shape).size == c.ndim
+    derivative = _poly_der(c)
+    for j in range(c.shape[1]):
+        for k in range(c.shape[2]):
+            np.testing.assert_allclose(
+                actual=derivative[:, j, k], desired=np.polyder(c[:, j, k])
+            )
+
+
+@pytest.mark.unit
+def test_poly_val():
+    """Test vectorized computation of polynomial evaluation."""
+
+    def test(x, c):
+        val = _poly_val(x=x, c=c)
+        if val.ndim != max(x.ndim, c.ndim - 1):
+            raise ValueError(f"Incompatible shapes {x.shape} and {c.shape}.")
+        for index in np.ndindex(c.shape[1:]):
+            idx = (..., *index)
+            np.testing.assert_allclose(
+                actual=val[idx],
+                desired=np.poly1d(c[idx])(x[idx]),
+                err_msg=f"Failed with shapes {x.shape} and {c.shape}.",
+            )
+
+    quartic = 5
+    c = np.arange(-60, 60).reshape(quartic, 3, -1) * np.pi
+    # make sure broadcasting won't hide error in implementation
+    assert np.unique(c.shape).size == c.ndim
+    x = np.linspace(0, 20, c.shape[1] * c.shape[2]).reshape(c.shape[1], c.shape[2])
+    test(x, c)
+
+    x = np.stack([x, x * 2], axis=0)
+    x = np.stack([x, x * 2, x * 3, x * 4], axis=0)
+    # make sure broadcasting won't hide error in implementation
+    assert np.unique(x.shape).size == x.ndim
+    assert c.shape[1:] == x.shape[x.ndim - (c.ndim - 1) :]
+    assert np.unique((c.shape[0],) + x.shape[c.ndim - 1 :]).size == x.ndim - 1
+    test(x, c)
+
+
+@pytest.mark.unit
+def test_get_extrema():
+    """Test computation of extrema of |B|."""
+    start = -np.pi
+    end = -2 * start
+    k = np.linspace(start, end, 5)
+    B = CubicHermiteSpline(
+        k, np.cos(k) + 2 * np.sin(-2 * k), -np.sin(k) - 4 * np.cos(-2 * k)
+    )
+    B_z_ra = B.derivative()
+    extrema, B_extrema = _get_extrema(k, B.c, B_z_ra.c)
+    extrema, B_extrema = map(_filter_not_nan, (extrema, B_extrema))
+    idx = np.argsort(extrema)
+
+    extrema_scipy = np.sort(B_z_ra.roots(extrapolate=False))
+    B_extrema_scipy = B(extrema_scipy)
+    assert extrema.size == extrema_scipy.size
+    np.testing.assert_allclose(extrema[idx], extrema_scipy)
+    np.testing.assert_allclose(B_extrema[idx], B_extrema_scipy)
+
+
+@pytest.mark.unit
+def test_composite_linspace():
+    """Test this utility function useful for Newton-Cotes integration over pitch."""
+    B_min_tz = np.array([0.1, 0.2])
+    B_max_tz = np.array([1, 3])
+    breaks = np.linspace(B_min_tz, B_max_tz, num=5)
+    b = _composite_linspace(breaks, num=3)
+    for i in range(breaks.shape[0]):
+        for j in range(breaks.shape[1]):
+            assert only1(np.isclose(breaks[i, j], b[:, j]).tolist())
+
+
+@pytest.mark.unit
+def test_bounce_points():
```

Comment:
this might be better as a class to group these, so that you can run them individually